

问题描述

benchmarksql跑TPCC几次之后，报错如下：



![微信图片_20210417195324](F:\学习笔记\postgresql\pg\postgresql\PG流复制总结\微信图片_20210417195324.jpg)



分析：

可能是流复制同步的操作与benchmarksql执行的sql冲突导致。

备库的postgresql.conf增加下面配置

hot_standby_feedback=on
max_standby_streaming_delay=300s

开启从节点流复制反馈，

从节点的查询最长运行时间为300s，超过该值就杀死查询。





参数解释：

hot_standby_feedback

  指定一个热后备机是否将会向主服务器或上游后备机发送有关于后备机上当前正被执行
的查询的反馈。这个参数可以被用来排除由于记录清除导致的查询取消，但是可能导致
在主服务器上用于某些负载的数据库膨胀。反馈消息的发送频度不会高于每
个wal_receiver_status_interval周期发送一次。默认值是off。这个参数只能
在postgresql.conf文件中或在服务器命令行上设置。  

max_standby_streaming_delay

  当热后备机处于活动状态时，这个参数决定取消那些与即将应用的 WAL 项冲突的后备机
查询之前，后备服务器应该等待多久，如第 26.5.2 节 中所述。 当 WAL 数据正在通过
流复制被接收时，max_standby_streaming_delay可以应用。 如果指定值时没有单位，
则以毫秒为单位。默认值是 30 秒。值 -1 允许后备机一直等到冲突查询结束。 这个参
数只能在postgresql.conf文件中或在服务器命令行上设置。
注意，max_standby_streaming_delay与取消之前一个查询能够运行的最长时间不同；它
表示在从主服务器接收到 WAL 数据并立刻应用它能够被允许的最长总时间。因此，如果
一个查询导致了显著的延迟，后续冲突查询将只有更少的时间，直到后备服务器再次赶
上进度。



#### 模拟查询冲突



1.主库

create table test1(id int , info text, crt_time timestamp);
insert into test1 select 1,md5(random()::text),now() from generate_series(1,1000);

2.备库

begin;
select count(*) from test1;

3.主库

delete from test1 ;

4.备库等待2-3min，再次执行

select count(*) from test1;

发现连接已经中断并恢复

FATAL:  terminating connection due to conflict with recovery
外地关闭了联接
	这种现象通常意味着服务器在处理请求之前
或者正在处理请求的时候意外中止
与服务器的连接已断开，正在试图重置: 完成。

后台日志内容：

2021-04-24 10:35:07.217 EDT [4255] FATAL:  terminating connection due to conflict with recovery
2021-04-24 10:35:07.217 EDT [4255] DETAIL:  User was holding a relation lock for too long.
2021-04-24 10:35:07.217 EDT [4255] HINT:  In a moment you should be able to reconnect to the database and repeat your command.



查看后台进程：

终止连接之前：

[root@node2 log]# ps -ef|grep postgres
txadmin   4184     1  0 10:23 ?        00:00:00 /usr/txdb4.0/bin/postgres
txadmin   4185  4184  0 10:23 ?        00:00:00 postgres: logger   
txadmin   4186  4184  0 10:23 ?        00:00:00 postgres: startup   recovering 0000000300000001000000F2
txadmin   4193  4184  0 10:23 ?        00:00:00 postgres: checkpointer   
txadmin   4194  4184  0 10:23 ?        00:00:00 postgres: background writer  
txadmin   4195  4184  0 10:23 ?        00:00:00 postgres: stats collector  
txadmin   4196  4184  0 10:23 ?        00:00:01 postgres: walreceiver   streaming 1/F2AD7D50
**txadmin   4255  4184  0 10:29 ?        00:00:00 postgres: txadmin txdb_default [local] idle in transaction**
txadmin   4296  3126  0 10:33 pts/1    00:00:00 tailf postgresql-Sat.log
root      4350  4307  0 10:34 pts/2    00:00:00 grep --color=auto postgres

终止之后：

[root@node2 log]# ps -ef|grep postgres
txadmin   4184     1  0 10:23 ?        00:00:00 /usr/txdb4.0/bin/postgres
txadmin   4185  4184  0 10:23 ?        00:00:00 postgres: logger   
txadmin   4186  4184  0 10:23 ?        00:00:00 postgres: startup   recovering 0000000300000001000000F2
txadmin   4193  4184  0 10:23 ?        00:00:00 postgres: checkpointer   
txadmin   4194  4184  0 10:23 ?        00:00:00 postgres: background writer  
txadmin   4195  4184  0 10:23 ?        00:00:00 postgres: stats collector  
txadmin   4196  4184  0 10:23 ?        00:00:01 postgres: walreceiver   streaming 1/F2ADB648
txadmin   4351  3126  0 10:34 pts/1    00:00:00 tailf postgresql-Sat.log
root      4353  4307  0 10:35 pts/2    00:00:00 grep --color=auto postgres



解释：主机在删除表test数据后，备库上的连接仍持有表test的锁，导致备库在replay该WAL日志时和已有的事务发生冲突，数据库终止了已有的连接会话。